ecs-agent/netlib/network_builder.go

```diff
@@ -1,46 +1,187 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
 package netlib
 
 import (
 	"context"
 
 	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
-	"github.com/aws/amazon-ecs-agent/ecs-agent/netlib/model/ecscni"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/data"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/ecs_client/model/ecs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+	netlibdata "github.com/aws/amazon-ecs-agent/ecs-agent/netlib/data"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/netlib/model/status"
 	"github.com/aws/amazon-ecs-agent/ecs-agent/netlib/model/tasknetworkconfig"
 	"github.com/aws/amazon-ecs-agent/ecs-agent/netlib/platform"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/volume"
 
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 )
 
 type NetworkBuilder interface {
 	BuildTaskNetworkConfiguration(taskID string, taskPayload *ecsacs.Task) (*tasknetworkconfig.TaskNetworkConfig, error)
 
-	Start(ctx context.Context, taskNetConfig *tasknetworkconfig.TaskNetworkConfig) error
+	Start(ctx context.Context, mode string, taskID string, netNS *tasknetworkconfig.NetworkNamespace) error
 
-	Stop(ctx context.Context, taskNetConfig *tasknetworkconfig.TaskNetworkConfig) error
+	Stop(ctx context.Context, mode string, taskID string, netNS *tasknetworkconfig.NetworkNamespace) error
 }
 
 type networkBuilder struct {
-	platformAPI platform.API
+	platformAPI    platform.API
+	metricsFactory metrics.EntryFactory
+	networkDAO     netlibdata.NetworkDataClient
 }
 
-func NewNetworkBuilder(platformString string) (NetworkBuilder, error) {
-	pAPI, err := platform.NewPlatform(platformString, ecscni.NewNetNSUtil())
+func NewNetworkBuilder(
+	platformString string,
+	metricsFactory metrics.EntryFactory,
+	volumeAccessor volume.VolumeAccessor,
```

Comment:
The VolumeAccessor can be used by docker agent as well. Although there is another reason to include it in this PR. If we were to implement containerd agent's VolumeAccessor in the shared library, we would have to import several other code fragments which are not unified at this point, which would cascade in further more changes into the shared library needlessly. Hence having this interface helps us to prevent that and postpone the unification of those to the time when ECS/EC2 adopts the containerd agent. I do not see a reason for this to be a one way door decision. We can always get rid of this interface without any compatibility issues and include the operation straight into the platform layer if required.

Reply:
ack